src/alpaka/AlpakaCore/alpaka/ScopedContext.cc

```diff
@@ -0,0 +1,78 @@
+#include "AlpakaCore/alpakaConfig.h"
+#include "AlpakaCore/ScopedContext.h"
+
+namespace {
+  struct CallbackData {
+    edm::WaitingTaskWithArenaHolder holder;
+    int device;
+  };
+}  // namespace
+
+namespace cms::alpakatools::ALPAKA_ACCELERATOR_NAMESPACE {
+
+  namespace impl {
+    void ScopedContextGetterBase::synchronizeStreams(int dataDevice,
+                                                     ::ALPAKA_ACCELERATOR_NAMESPACE::Queue& dataStream,
+                                                     bool available,
+                                                     alpaka::Event<::ALPAKA_ACCELERATOR_NAMESPACE::Queue> dataEvent) {
+      if (dataDevice != device()) {
+        // Eventually replace with prefetch to current device (assuming unified memory works)
+        // If we won't go to unified memory, need to figure out something else...
+        throw std::runtime_error("Handling data from multiple devices is not yet supported");
+      }
+
+      if (dataStream != stream()) {
+        // Different streams, need to synchronize
+        if (not available) {
+          // Event not yet occurred, so need to add synchronization
+          // here. Sychronization is done by making the CUDA stream to
+          // wait for an event, so all subsequent work in the stream
+          // will run only after the event has "occurred" (i.e. data
+          // product became available).
+          alpaka::wait(stream(), dataEvent);
+        }
+      }
+    }
+
+    void ScopedContextHolderHelper::enqueueCallback(int device, ::ALPAKA_ACCELERATOR_NAMESPACE::Queue& stream) {
+      alpaka::enqueue(stream, [this, device]() {
+        auto data = new CallbackData{waitingTaskHolder_, device};
+        std::unique_ptr<CallbackData> guard{reinterpret_cast<CallbackData*>(data)};
+        edm::WaitingTaskWithArenaHolder& waitingTaskHolder = guard->holder;
+        int device2 = guard->device;
+        waitingTaskHolder.doneWaiting(nullptr);
+      });
```

Comment:
How does the `alpaka::enqueue()` propagate errors from asynchronous processing? Is it like `cudaHostFunc()` that doesn't call the function in case of errors, an exception pops up elsewhere, or something else?